feat: expand `processInputStep` and integrate `prepareStep` as processor

CalebBarnes opened the pull request:
## Description

Expands the `processInputStep` processor method and integrates `prepareStep` to run through the standard processor pipeline. This PR enables per-step configuration changes during the agentic loop.

### Key Changes

1. **prepareStep as a processor**: `prepareStep` is now internally wrapped as `PrepareStepProcessor` and runs through the standard `processInputStep` pipeline.

2. **Expanded ProcessInputStepArgs**: Processors now receive much more context:
   - `messages` - Current messages snapshot (read-only)
   - `systemMessages` - Current system messages (read-only)
   - `messageList` - Full MessageList instance for mutations
   - `model` - Current model
   - `tools` - Current tools available for this step
   - `toolChoice` - Current tool choice setting
   - `activeTools` - Currently active tools
   - `steps` - Results from previous steps
   - `providerOptions` - Current provider options
   - `modelSettings` - Current model settings
   - `structuredOutput` - Structured output configuration

3. **New ProcessInputStepResult type**: Processors can now return:
   - `model` - Change model for this step (with model config resolution)
   - `tools` - Replace or add tools (dynamically add tools mid-conversation)
   - `toolChoice` - Change tool choice
   - `activeTools` - Filter active tools  
   - `messages` - Replace messages
   - `messageList` - Return same messageList (indicates mutations)
   - `systemMessages` - Replace all system messages
   - `providerOptions` - Change provider options
   - `modelSettings` - Change model settings
   - `structuredOutput` - Change structured output configuration

4. **Processor chaining**: Each processor receives accumulated state from previous processors. If Processor 1 returns `{ model: modelA }`, Processor 2 receives `model: modelA` in its args (not the original).

5. **System message isolation**: System messages are reset to their original values at the start of each step. Modifications made in `prepareStep`/`processInputStep` only affect the current step.

6. **Tools passed to toolCallStep**: Tools modified in `prepareStep`/`processInputStep` are now correctly passed through to the tool execution step, so dynamically added tools can be executed.

7. **Validation and error handling**:
   - Validates that returned MessageList is the same instance passed in
   - Validates that both `messages` and `messageList` aren't returned together
   - Validates that v1 models aren't returned (only v2 supported)
   - Resolves model config strings (e.g., `"openai/gpt-4o"`) to model instances

8. **Improved observability**: Processor spans include all serializable input args and output results.

9. **Type improvements**: 
   - Added `void | Promise<void>` to `Processor.processInputStep` return type
   - Generic `TOOLS` type flows through `ProcessInputStepArgs` and `ProcessInputStepResult`
   - New `RunProcessInputStepArgs` and `RunProcessInputStepResult` types for the runner

### Breaking Changes

- **prepareStep messages format changed**: `prepareStep` used to receive `messages` formatted as AI SDK v5 model messages. Now that it's unified with `processInputStep`, messages are in `MastraDBMessage` format.
  - **Migration**: Use `messageList.get.all.aiV5.model()` if you need the old format
  - Example:
    ```typescript
    prepareStep: async ({ messageList }) => {
      const aiSdkMessages = messageList.get.all.aiV5.model();
      // ... use aiSdkMessages
    };
    ```

### Example Usage

```typescript
// Dynamic model switching per step
agent.generate('Hello', {
  prepareStep: async ({ stepNumber }) => {
    if (stepNumber === 0) return { model: 'openai/gpt-4o-mini' };
    return { model: 'openai/gpt-4o' };
  }
});

// Dynamically add tools based on conversation context
agent.generate('Hello', {
  prepareStep: async ({ stepNumber, tools }) => {
    if (stepNumber === 1) {
      return {
        tools: {
          ...tools,
          dynamicTool: tool({
            inputSchema: z.object({ query: z.string() }),
            execute: async () => 'dynamic result',
          }),
        },
        toolChoice: { type: 'tool', toolName: 'dynamicTool' },
      };
    }
  }
});

// Per-step system instructions (isolated to each step)
agent.generate('Hello', {
  prepareStep: async ({ stepNumber, systemMessages }) => {
    return {
      systemMessages: [
        ...systemMessages,
        { role: 'system', content: `Step ${stepNumber} instructions` }
      ]
    };
  }
});

// Provider options per step
agent.generate('Hello', {
  prepareStep: async ({ stepNumber }) => {
    if (stepNumber === 0) {
      return {
        providerOptions: {
          anthropic: { cacheControl: { type: 'ephemeral' } }
        }
      };
    }
  }
});

// Model settings per step
agent.generate('Hello', {
  prepareStep: async ({ stepNumber, modelSettings }) => {
    if (stepNumber > 3) {
      return {
        modelSettings: { ...modelSettings, maxTokens: 500 }
      };
    }
  }
});

// Disable tools after certain steps
agent.generate('Hello', {
  prepareStep: async ({ stepNumber }) => {
    if (stepNumber > 5) {
      return { toolChoice: 'none' };
    }
  }
});

// Access previous step results
agent.generate('Hello', {
  prepareStep: async ({ steps }) => {
    if (steps.length > 0 && steps[steps.length - 1].finishReason === 'tool-calls') {
      // React to tool calls from previous step
    }
  }
});
```

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Documentation updated (migration guide, processor docs, reference docs)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * prepareStep now runs in the processor pipeline, letting processors modify per-step model, tools, toolChoice, activeTools, messages, systemMessages, providerOptions, modelSettings, and structuredOutput.
  * Processor chaining accumulates and passes state across steps.
  * System messages are reset at each step for isolation.

* **Breaking Changes**
  * prepareStep message format changed to MastraDBMessage; use messageList.get.all.aiV5.model() to obtain the legacy format.

* **Documentation**
  * Expanded processor docs, migration guide, examples, and reference material.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

CalebBarnes committed:
prepareStep support model id + config obj / messages in mastra db format / pass messageList to prepareStep

CalebBarnes committed:
add assistant and tool messages as response source

CalebBarnes committed:
prepareStep test updating

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
wip preparestep input step processor

CalebBarnes committed:
wip prepare step processor

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
wip prepare step processor

CalebBarnes committed:
more options

CalebBarnes committed:
test

CalebBarnes committed:
prune

CalebBarnes committed:
lint fix

CalebBarnes committed:
rename prune -> cleanup

CalebBarnes committed:
allow void returns

CalebBarnes committed:
allow void return

CalebBarnes committed:
store initialSystemMessages before llm execution and restore them before running prepareStep/processInputStep

CalebBarnes committed:
add prepareStep test: system message modifications in prepareStep should not persist across steps

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
add system messages

CalebBarnes committed:
plan

CalebBarnes committed:
add support for modifying providerOptions via prepareStep

- Pass providerOptions to processInputStep and allow returning modified value
- Add stepProviderOptions variable in llm-execution-step to track modifications
- Apply providerOptions from processInputStepResult to the model execution
- Add test to verify providerOptions can be set via prepareStep callback

CalebBarnes committed:
messageListPromptArgs use stepModel

CalebBarnes committed:
add modelSettings support and fix processor chaining

- Wire up modelSettings in llm-execution-step (stepModelSettings variable)
- Fix processor chaining: each processor now receives accumulated values from
  previous processors (model, toolChoice, activeTools, providerOptions, modelSettings)
- Add tests for processor chaining (model and providerOptions)
- Add test for modelSettings modification via prepareStep

CalebBarnes committed:
fix: include messageList in runProcessInputStep result

CalebBarnes committed:
cleanup command delete .mastra dir

CalebBarnes committed:
refactor: simplify processor chaining and improve span logging

- Remove separate current* tracking variables, use stepInputResult directly
- Use current span as parent for input step processor spans (not AGENT_RUN)
- Expand span input to include all serializable processor args
- Expand span output to include result details instead of just messages

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
Add tests for messages/systemMessages/messageList modifications in processInputStep

- Add tests for messages modification via return { messages: [...] }
- Add tests for messages chaining through multiple processors
- Add tests for systemMessages modification via return { systemMessages: [...] }
- Add tests for systemMessages chaining through multiple processors
- Add tests for messageList direct mutations
- Update toolChoice types to allow ToolChoice<any> for flexibility
- Add cast in llm-execution-step for toolChoice assignment
- Update v1 migration guide (agent.mdx) with prepareStep messages format breaking change

CalebBarnes committed:
docs: add processInputStep API docs and test coverage

- Add complete ProcessInputStepArgs/Result docs to processor reference
- Add processor chaining and system message isolation sections
- Add Processor Interface to sidebar navigation
- Add tests for abort, validation, and error handling
- Update PLAN.md to reflect completed test coverage

CalebBarnes committed:
Add structuredOutput support to processInputStep

- Add structuredOutput to ProcessInputStepArgs and ProcessInputStepResult
- Remove OUTPUT generic from processor types (schema type is OutputSchema base type)
- Add structuredOutput chaining through processor pipeline
- Add tests for structuredOutput modification and Zod schema extension
- Fix prepareStep test to use MastraDBMessage format

CalebBarnes committed:
refactor: move activeTools to top-level ExecutionProps

Move activeTools from nested options to top-level in ExecutionProps,
alongside tools and toolChoice for consistency. Also pass tools to
processInputStep and handle tools replacement in the processor chain.

CalebBarnes committed:
feat: add tools support to processInputStep

Processors can now modify tools at each step of the agentic loop:
- Pass tools to processors via ProcessInputStepArgs
- Processors can return { tools } to replace tools for the step
- Tools chain through multiple processors like other settings
- Use spread to merge: { tools: { ...tools, newTool } }

CalebBarnes committed:
fix: cast defaultOptions type for OUTPUT generic compatibility

CalebBarnes committed:
refactor: move activeTools from LoopConfig to LoopOptions

Move activeTools to top-level LoopOptions alongside tools and toolChoice
for consistency. Update all usages to reference the new location.

CalebBarnes committed:
dont cast as unknonw

CalebBarnes committed:
delete workflowResourceId

CalebBarnes committed:
remove empty options

CalebBarnes committed:
comment

CalebBarnes committed:
add test for prepareStep tools modification

CalebBarnes committed:
agent test use mock model

DanielSLew reviewed (approved):
Looks good! Just a nit

CalebBarnes committed:
agent prepareStep tools tests

CalebBarnes committed:
pass tools modified in prepareStep/processInputStep to toolCallStep

- Add tools to LLMIterationOutput and toolCallInputSchema
- Pass tools from llmExecutionStep through the workflow to toolCallStep
- toolCallStep uses inputData.tools with fallback to closure tools
- Update docs with tools and structuredOutput in processInputStep
- Update processor-interface.mdx execution order to clarify prepareStep runs as part of processInputStep pipeline

CalebBarnes committed:
remove console logs / update test

CalebBarnes committed:
remove commented code and unnecessary type cast

CalebBarnes committed:
spread stepInput to span output

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
delete md file

CalebBarnes committed:
add prepareStep structuredOutput tests

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
remove console logs

CalebBarnes committed:
fix: store prepareStep tools in _internal to survive workflow suspend/resume

Tools modified via prepareStep/processInputStep were being passed through
workflow inputData, but when the workflow suspends (e.g., for tool approval)
and resumes, the inputData gets serialized. Since functions cannot survive
JSON serialization, the tool execute functions were lost on resume.

Solution: Store modified tools in _internal.stepTools instead. The _internal
object is a mutable closure that persists across workflow steps without
being serialized, preserving the execute functions.

Also fixed test utilities to use deterministic message IDs for snapshots.

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
add changeset

CalebBarnes committed:
add changesetremove tools type from LLMIterationOutput

CalebBarnes committed:
Merge branch 'main' into fix/prepare-step-message-list

CalebBarnes committed:
Use currentStep.tools in processOutputStream to honor per-step tool overrides
